build-tools/packages/build-cli/src/commands/release/setDtsTypes.ts

```diff
@@ -0,0 +1,182 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import { Flags } from "@oclif/core";
+import { Package, updatePackageJsonFile, PackageJson } from "@fluidframework/build-tools";
+import { PackageCommand } from "../../BasePackageCommand";
+import { ExtractorConfig } from "@microsoft/api-extractor";
+import { CommandLogger } from "../../logging";
+import path from "node:path";
+import { strict as assert } from "node:assert";
+import * as fs from "fs";
+
+/**
+ * Represents a list of package categorized into two arrays
+ */
+interface PackageTypesList {
+	/**
+	 * An array of strings containing package names that have been updated.
+	 */
+	packagesUpdated: string[];
+	/**
+	 * An array of strings containing package names that have not been updated.
+	 */
+	packagesNotUpdated: string[];
+}
+
+const knownDtsKinds = ["alpha", "beta", "public", "untrimmed"] as const;
+
+type DtsKind = typeof knownDtsKinds[number];
+
+function isDtsKind(str: string | undefined): str is DtsKind {
+	return str === undefined ? false : knownDtsKinds.includes(str as any);
+}
+
+export default class SetReleaseTagPublishingCommand extends PackageCommand<
+	typeof SetReleaseTagPublishingCommand
+> {
+	static readonly description =
+		"Updates the .d.ts types that are included in the `types` field in package.json. This command is intended to be used in CI only.";
+
+	static readonly enableJsonFlag = true;
+
+	static readonly flags = {
+		types: Flags.custom<DtsKind>({
+			description: "Which .d.ts types to include in the published package.",
+			required: true,
+			parse: async (input) => {
+				if (isDtsKind(input)) {
+					return input;
+				}
+				throw new Error(`Invalid release type: ${input}`);
+			},
+		})(),
+		...PackageCommand.flags,
+	};
+
+	private readonly packageList: PackageTypesList = {
+		packagesNotUpdated: [],
+		packagesUpdated: [],
+	};
+
+	protected async processPackage(pkg: Package): Promise<void> {
+		updatePackageJsonFile(pkg.directory, (json) => {
+			if (json.types !== undefined && json.typings !== undefined) {
+				throw new Error(
+					"Both 'types' and 'typings' fields are defined in the package.json. Only one should be used.",
+				);
+			}
+
+			const types: string | undefined = json.types ?? json.typings;
+
+			if (types === undefined) {
+				this.verbose("Neither 'types' nor 'typings' field is defined in the package.json.");
+				return;
+			}
+
+			const configOptions = ExtractorConfig.tryLoadForFolder({
+				startingFolder: pkg.directory,
+			});
+			if (configOptions === undefined) {
```

Comment:
I think this check should be done first (before calling updatePackageJsonFile) so packages not using API extractor don't get their package.json file rewritten.

Also, then I think it would be better to error in the case where there are no types in the package.json instead of just warning, since all out packages which use API-Extractor should have types.